open_spiel/games/checkers.cc

```diff
@@ -0,0 +1,546 @@
+              action_values[1] = column;                              // Initial column value of player piece
+              action_values[2] = direction;                           // Direction of move for player piece
+              action_values[3] = MoveType::kNormal;                   // Type of move
+              action_values[4] = PieceType::kMan;                     // Type of captured piece if any. kMan by default
+              action_values[5] = StateToPiece(BoardAt(row, column));  // Type of player piece
+              move_list.push_back(
+                  RankActionMixedBase(action_bases, action_values));
+            } else if (adjacent_state == opponent_state || adjacent_state == opponent_state_crowned) {
+              int jumping_row = adjacent_row + kDirRowOffsets[direction];
+              int jumping_column = adjacent_column + kDirColumnOffsets[direction];
+              if (InBounds(jumping_row, jumping_column) && BoardAt(jumping_row, jumping_column) == CellState::kEmpty ) {
+                action_values[0] = row;
+                action_values[1] = column;
+                action_values[2] = direction;
+                action_values[3] = MoveType::kCapture;
+                action_values[4] = StateToPiece(adjacent_state);
+                action_values[5] = StateToPiece(BoardAt(row, column));
+                capture_move_list.push_back(
+                    RankActionMixedBase(action_bases, action_values));
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  // If capture moves are possible, it's mandatory to play them.
+  if (!capture_move_list.empty()) {
+    if (multiple_jump_piece_ > 0) {
+      int multiple_jump_piece_row = multiple_jump_piece_ / rows_;
+      int multiple_jump_piece_column = multiple_jump_piece_ % rows_;
+      std::vector<Action> multiple_move_list;
+      for (Action action: capture_move_list) {
+        std::vector<int> move = UnrankActionMixedBase(action, {rows_, columns_, kNumDirections, kNumMoveType, kNumPieceType, kNumPieceType});
+        if (move[0] == multiple_jump_piece_row && move[1] == multiple_jump_piece_column) {
+          multiple_move_list.push_back(action);
+        }
+      }
+      SPIEL_CHECK_GT(multiple_move_list.size(), 0);
+      return multiple_move_list;
+    }
+    return capture_move_list;
+  }
+  return move_list;
+}
+
+bool CheckersState::InBounds(int row, int column) const {
+  return (row >= 0 && row < rows_ && column >= 0 && column < columns_);
+}
+
+std::string CheckersState::ToString() const {
+  std::string result = "";
+  for (int r = 0; r < rows_; r++) {
+    // Ensure the row labels are aligned.
+    if (rows_ - r < 10 && rows_ >= 10) {
+      absl::StrAppend(&result, " ");
+    }
+    absl::StrAppend(&result, RowLabel(rows_, r));
+
+    for (int c = 0; c < columns_; c++) {
+      absl::StrAppend(&result, StateToString(BoardAt(r, c)));
```

Comment:
Is it worth distinguishing between empty black/white squares, as in one that pieces can be on vs ones that are inaccessible? Maybe with dot vs space?

Reply:
I didn't understand this correctly. Can you elaborate?